src/rust/inetstack/protocols/udp/peer.rs

```diff
@@ -234,126 +263,125 @@ impl<const N: usize> UdpPeer<N> {
             addr.set_port(new_port);
         }
 
-        // Bind endpoint and create a receiver-side shared queue.
-        queue.set_addr(addr);
-        queue.set_recv_queue(SharedQueue::<SharedQueueSlot<DemiBuffer>>::new(RECV_QUEUE_MAX_SIZE));
-        self.bound.insert(addr, qd);
+        queue.bind(addr)?;
         Ok(())
     }
 
     /// Closes a UDP socket.
     pub fn do_close(&mut self, qd: QDesc) -> Result<(), Fail> {
         #[cfg(feature = "profiler")]
         timer!("udp::close");
-        let mut qtable: RefMut<IoQueueTable> = self.qtable.borrow_mut();
-        // Lookup associated endpoint.
-        let queue: UdpQueue = qtable.free::<UdpQueue>(&qd)?;
-        match queue.get_addr() {
-            Ok(addr) => {
-                self.bound.remove(&addr);
-                Ok(())
-            },
-            Err(e) => Err(e),
-        }
+        let queue: SharedUdpQueue<N> = self.qtable.borrow_mut().free::<SharedUdpQueue<N>>(&qd)?;
+        queue.close()?;
+        Ok(())
     }
 
     /// Pushes data to a remote UDP peer.
-    pub fn do_pushto(&self, qd: QDesc, data: DemiBuffer, remote: SocketAddrV4) -> Result<(), Fail> {
+    pub fn do_pushto(&mut self, qd: QDesc, data: DemiBuffer, remote: SocketAddrV4) -> Result<(), Fail> {
         #[cfg(feature = "profiler")]
         timer!("udp::pushto");
-        let qtable: Ref<IoQueueTable> = self.qtable.borrow();
         // Lookup associated endpoint.
-        let queue: &UdpQueue = qtable.get::<UdpQueue>(&qd)?;
-        let local: SocketAddrV4 = queue.get_addr()?;
-
+        let queue: SharedUdpQueue<N> = self.get_queue(&qd)?;
+        // What happens if not bound?
```

Comment:
Pushing to an unbound socket is fine. This is the main reason why `pushto` exists.